poc/flp.py

```diff
@@ -1,17 +1,18 @@
 """Fully linear proof (FLP) systems."""
 
+from typing import Any
+
 import field
 from common import vec_add, vec_sub
-from field import Field
 
 
 class Flp:
     """The base class for FLPs."""
 
     # Generic paraemters
-    Measurement = None
-    AggResult = None
-    Field: field.Field = None
+    Measurement: Any = None
+    AggResult: Any = None
+    Field: Any = None
```

Comment:
Reviewer note: I tried this because I thought it would be more natural:

```suggestion
    Field: type[field.Field]
```

This doesn't work: ttps://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases

Another alternative is to make this generic. For instance we could do:
```python
F = TypeVar('F', bound=field.Field)
class Flp(Generic[F]):
   field: F

class FlpField64(Flp[Field64]):
   field = Field64
```

Generic may in fact be the way to go in the long run, but it requires more refactoring.

Note that we still we don't get to use `F` as a type: it exists only to support type hints. Thus it is still necessary to define `field` as an attribute.

Also, note that as of python 3.12, there is some syntactic sugar that makes this look nicer:
```python
class Flp[F: Field]:
   field: F
```

Reply:
I think the problem here is fundamentally one of partial initialization, due to how we use class factories. Instead of `Any`, we could hypothetically use something like `Union[type[field.Field], NoneType]`. Note that `None` would still be a problem even if we had generic type variables, as the type hint would then be `Union[type[F], NoneType]`.

Moreover, the final class we construct from a class factory only really exists at runtime, so static analysis-based type checking will run into some fundamental limitations with this pattern.

Reply:
This is indeed messy. I find I often end up in trying to over-engineer things like this in Python. I think to address this issue I want to more or less follow what we've already done. Using `Any` for the "abstract base class" solves things without too many changes.